### PR TITLE
[bugfix] Fix minimum values for range widgets

### PR DESCRIPTION
**Pull request by elpaso**
Fixes #17878 QGIS 3 Vector Layer Fields Garbled when Clicking the Toggle Editing Icon

### CODE REVIEW

**Comment by dgoedkoop** [comment]
I have just tested this, and it fixes the bug of #17878 only partly. The data are still altered (the and layer is marked dirty) when just pressing the 'enable editing' button in the attribute table, which I think shouldn't happen.

**Comment by nyalldawson** [comment]
I find it hard to follow the logic here... Can we get some unit tests protecting this?

**Comment by elpaso** [comment]
@dgoedkoop thanks for testing, I know it's a partial fix because the bug has not a single root, as I said in the bug comments there is another problem related to precision (number of digits) in the range widget and I believe it should be addressed separately.

**Comment by dgoedkoop** [comment]
@elpaso yes. I can see the following four themes:
- Value set to NULL unexpectedly
- Precision of the range widget
- Sometimes the save() function for cells in the attribute table is maybe called too often, e.g. when enabling editing or when deleting features
- Something seems to be off with the bookkeeping in the attribute table, as evident from the facts that
   - the bug doesn't happen again when you disable and re-enable editing without closing the form
   - the bug does sometimes happen again when you re-sort the table and delete some rows
   - if you delete some rows, and your selection included the first row in the table, twice as many rows disappear from the attribute table (but not from the actual data).

**Comment by elpaso** [comment]
> if you delete some rows, and your selection included the first row in the table, twice as many rows disappear from the attribute table (but not from the actual data).
@luipir ^ didn't you work on something similar in Essen?

**Comment by dgoedkoop** [comment]
> if you delete some rows, and your selection included the first row in the table, twice as many rows disappear from the attribute table (but not from the actual data).

I've found the cause in a 87 levels deep stack trace

- The attribute _table_ sends a message to the qgisapp, to delete the features
- qgisapp deletes the features and emits a `endEditCommand()`
- The attribute _table_ removes the rows from the table
- The feature list of the attribute _form_ is hooked into the endRemoveRows() event of the _table_
- Which, as per 3) of my previous comment, checks if any feature is selected for editing in the attribute _form_, which is not the case, because it has just been deleted. So it selects a new feature
- `mFeatureList_aboutToChangeEditSelection` is triggered, which for some reason calls `save()`, which calls `saveEdits()`
- `saveEdits()` discovers that values have been changed (because of the precision error in the range widget), commits the changes, and calls `endEditCommand()` yet again
- The deletions are still in the queue from the previous `endEditCommand()`, which is still in progress. Hence, the row deletion event for the attribute _table_ is called again.

So again, if the range widget is fixed this problem will disappear. Yet I still have the feeling that this part of the code is not very robust.

**Comment by elpaso** [comment]
> So again, if the range widget is fixed this problem will disappear.

The problem is: what is the correct fix? I mean that the underlying widget (`QgsDoubleSpinBox` which is a subclass of `QDoubleSpinBox` needs a precision to work correctly, whatever precision we choose we will always end up with a "changed" value (even when the number of digits will not decrease).

> Yet I still have the feeling that this part of the code is not very robust.

I totally share your feeling, every time I work on the attribute table is a huge pain.

**Comment by elpaso** [comment]
Working on tests ...

**Comment by luipir** [comment]
@elpaso sorry for my late answer I erroneously remove the notification mail. Yes I remember something similar. @dgoedkoop found a possibile cause that depends on specific precision setting error. Is this  behaviour (duplication) present also with other data type?
